backend/src/main/kotlin/com/tngtech/apicenter/backend/domain/handler/ServiceHandler.kt

```diff
@@ -1,28 +1,46 @@
 package com.tngtech.apicenter.backend.domain.handler
 
+import com.tngtech.apicenter.backend.config.ApiCenterProperties
 import com.tngtech.apicenter.backend.domain.entity.ReleaseType
 import com.tngtech.apicenter.backend.domain.entity.ResultPage
-import com.tngtech.apicenter.backend.domain.entity.ServiceId
-import com.tngtech.apicenter.backend.domain.entity.Service
-import com.tngtech.apicenter.backend.domain.entity.Specification
+import com.tngtech.apicenter.backend.domain.entity.PermissionType
+import com.tngtech.apicenter.backend.connector.rest.security.JwtAuthenticationProvider
+import com.tngtech.apicenter.backend.connector.rest.service.RemoteServiceUpdater
+import com.tngtech.apicenter.backend.domain.entity.*
+import com.tngtech.apicenter.backend.domain.exceptions.PermissionDeniedException
 import com.tngtech.apicenter.backend.domain.exceptions.SpecificationConflictException
 import com.tngtech.apicenter.backend.domain.exceptions.SpecificationDuplicationException
+import com.tngtech.apicenter.backend.domain.service.PermissionsManager
 import com.tngtech.apicenter.backend.domain.service.ServicePersistor
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.stereotype.Component
 
 @Component
 class ServiceHandler @Autowired constructor(
-        private val servicePersistor: ServicePersistor
+        private val servicePersistor: ServicePersistor,
+        private val remoteServiceUpdater: RemoteServiceUpdater,
+        private val jwtAuthenticationProvider: JwtAuthenticationProvider,
+        private val permissionsManager: PermissionsManager,
+        private val apiCenterProperties: ApiCenterProperties
 ) {
 
-    fun addNewSpecification(specification: Specification, serviceId: ServiceId, fileUrl: String?) {
+    fun addNewSpecification(specification: Specification, serviceId: ServiceId, fileUrl: String?, isPublic: Boolean) {
         val service: Service? = servicePersistor.findOne(serviceId)
 
         if (service == null) {
             saveNewService(specification, serviceId, fileUrl)
-        } else {
+            val userId = jwtAuthenticationProvider.getCurrentUserId()
```

Comment:
How threadsafe is that?

Reply:
Sorry, I don't understand the question... what is the danger?